sharing permissions feature, take home test erasmo ramirez

eramirez405 opened the pull request.

eramirez405 commented:
### Solution
To enable users to share Projects, Reports, and Documents while assigning permission levels, I introduced a SharingPermissions table as a control layer. This table includes:

- resource_type – A string that can be 'project', 'document', or 'report'.
- access_level – Can be one of 'view', 'edit', or 'full access'.

Foreign keys:

- user_id
- creator_id
- project_id (if resource_type is 'project')
- document_id (if resource_type is 'document')
- report_id (if resource_type is 'report')

These constraints ensure that each row corresponds to exactly one type of shared resource.

This solution would allow users to be able to assign projects, documents and reports to other users, and also be able to query for projects, documents, reports assigned to a certain user, documents and reports. Hence creating logic in the server to only allow the edit, index, show controller actions to either the owner/current_user or a user that is accessing a shared resource having the correct access_level.

Example of projects payload with sharing_permissions included:

`[
    {
        "id": 3,
        "name": "Test user project",
        "description": "test description",
        "user_id": 5,
        "created_at": "2025-03-09T23:17:51.498Z",
        "updated_at": "2025-03-09T23:17:51.498Z",
        "sharing_permissions": [
            {
                "id": 4,
                "name": "Robert Paulson",
                "email": "robertpaulson@example.com",
                "avatar_url": "https://avatar.iran.liara.run/username?username=robert+paulson",
                "permission": "view"
            },
            {
                "id": 1,
                "name": "Tyler Durden",
                "email": "tylerdurden@example.com",
                "avatar_url": "https://avatar.iran.liara.run/username?username=tyler+durden",
                "permission": "view"
            }
        ],
        "url": "http://localhost:5000/projects/3.json"
    }
];`

Ideally I would just return the sharing_permissions to the owner, or users with full access, other than that for shared resources I would just return the access_level the user making the request has, but I struggled with formatting the response in rails, for this reason I used logic in the frontend to identify the access_level of the current_user. Secured the backend and only displayed the options on the frontend based on the permission.

### Technical Observations

- React Hot Reload Missing – Noticed that the React code doesn’t hot reload, which slows development.
- Formatting Issues – No .prettierrc file; PRs would be easier to review with consistent formatting. I disabled my vscode formatter, but some files got formatted before I realized.
- Unique Constraints – Initially added unique constraints in a migration, but the server crashed without clear feedback. Rolling back the migration resolved the issue.
- By default the feature to add a permission was intended to just type the email and proceed, but I changed the user interface to make it more intuitive a user can only type emails of existing users, if having more time I would create a component to stay in the @radix-ui scope, but for simplicity I went with my known friend react-select
- The user experience of add/update permission of a project from the homepage vs from the project specific page is different, the one in the homepage the ideal more dynamic, in the project specific page I had to close the modal to trigger a refresh on the assigned permissions, with more time the ideal solution would have been to move the logic around assigning permissions to a hook as there is repetitive logic in some places with pretty similar patterns. With that the UX would be consistent for both cases.
- I started working on the backend and chose the name sharing_permissions then saw on the frontend there was already a suggested property called sharedWith, I think sharedWith is a better name, with more time I would have made changes to use the sharedWith name instead.

### Final Thoughts

This was my first project using Ruby on Rails. My preparation for this take home test was [Ruby in Twenty Minutes](https://www.ruby-lang.org/en/documentation/quickstart/) and then [Getting Started with Rails](https://edgeguides.rubyonrails.org/getting_started.html), they allowed me to understand the soul of ruby on rails in matters of taking away the boilerplate and persisting standard best practices for most of the common server features.

I got help of chatgpt that help me with some methods that the getting started tutorial did not cover for (e.g., .includes(:project).map(&:project).compact) 

### My personal to do
_This is my personal list on how I start tackling the issue by aiming for small tasks—the ones I accomplished and the ones that are next in my soon-to-do scope._

- [x] Be able to create permission for project
- [x] Be able to retrieve permissions for a project
- [x] Create constraint against duplication of permission for project
- [x] Be able to create permissions for document
- [x] Be able to create permissions for report
- [x] Be able to delete permissions on projects
- [x] Be able to delete permissions on documents
- [x] Be able to delete permissions on reports
- [x] Projects can be viewed by owner and users with sharing permission of view, edit or full
- [x] Create frontend view for shared projects
- [x] Secure in the server sharing features to correct access levels
- [x] Show/Hide in front end sharing features to correct access level
- [x] Show/Hide add document/report for own/shared
- [ ] Show shared documents in the home page
- [ ] Show shared reports in the home page
- [ ] Secure in server on users with full access and edit can edit projects
- [ ] Secure in server on users with full access and edit can edit documents
- [ ] Secure in server on users with full access and edit can edit reports
- [ ] Display edit button only to owned resources or users with permission in shared resources
- [ ] Display the owner of shared resources in the home page
- [ ] Clean up frontend current user prop, retrieve from server the access_level of resources shared
- [ ] Create components to be shared related to document/report item display, available users (less lines of code, components more readable).
- [ ] Separate Project components into different files, for more readability and easier to test independent components if there is more abstraction.
- [ ] Explore how to add unit tests for the react frontend with the rails file structure
- [ ] Add tests for the backend new features

thiagosil commented on app/controllers/documents_controller.rb:
```diff
@@ -63,7 +63,30 @@ def destroy
     def set_project
-      @project = current_user.created_projects.find(params[:project_id]) if params[:project_id]
+      shared = params[:shared]
+      if shared.present? && shared == 'true'
```
You have a very similar logic on several controllers, how would you avoid that?

eramirez405 replied:
Rails provides a feature called before_action, which allows us to extract shared logic from multiple actions into reusable methods, keeping our controllers DRY (Don't Repeat Yourself). Given more time to refine the code, I would identify the common logic across controllers and refactor it into before actions. This would improve maintainability and adherence to Rails' best practices.

thiagosil commented on app/javascript/app/api/useDeleteShareResource.ts:
```diff
@@ -0,0 +1,30 @@
+
+
+
+export default function useDeleteShareResource(resourceType: SharingPermission['resource_type'], cb?: () => void) {
```
What is cb? Prefer more informational names like callback

eramirez405 replied:
You're absolutely right, using more descriptive names improves readability. I used cb as a shorthand for 'callback', but I understand that explicit naming can make the code clearer for all readers.

thiagosil commented on app/javascript/app/api/useFetchProjects.ts:
```diff
@@ -1,9 +1,9 @@
 import { useQuery } from '@tanstack/react-query';
 import { getHeaders } from './helpers';
-export default function useFetchProjects() {
+export default function useFetchProjects(shared?: boolean) {
```
Shared is an optional parameter here, but it is required on useFetchProject and useFetchReports, we should have the same pattern for all the api hooks. What should be it, and why?

eramirez405 replied:
Thinking on separation of concerns and in that way making backend function get closer to single responsability, I would actually separate the endpoints for (resources being accessed by their owner) vs (resources being access by users with permission access to them), but as I had certain knoledge limitations on Rails, I decided to handle that separation in the frontend. But ideally we would have useFetchProjects and useFetchSharedProjects, in which 'shared' would not even be a parameter.

thiagosil commented on app/javascript/app/application.tsx:
```diff
@@ -30,14 +31,14 @@ const LandingPage = () => {
     if (isPending) return <Loader />;
 
-    return user ? <App /> : <WelcomePage />
+    return user ? <App user={user} /> : <WelcomePage />
```
Using this user on all components do not look very good. Do you know another way of solving that issue without adding this dependency like this?

eramirez405 replied:
Definitely agree with you, it's awful to see a property being passed on and on from parents to children, the reason I implented this way is what I described in the solution comment:

> Ideally I would just return the sharing_permissions to the owner, or users with full access, other than that for shared resources I would just return the access_level the user making the request has, but I struggled with formatting the response in rails, for this reason I used logic in the frontend to identify the access_level of the current_user. Secured the backend and only displayed the options on the frontend based on the permission.

So, having the knowledge to format the response on Rails appropriately I would just return the access_level to a shared resource within it's object properties.

Although if it becomes a necessity for the app to have a property being shared across multiple components, a global state management solution would be the way to go. React Context Api if just for the current user, Redux if more state will come into play.

thiagosil commented on app/controllers/sharing_permission_controller.rb:
```diff
@@ -0,0 +1,145 @@
+  def index
+    @sharing_permission = current_user.sharing_permissions.all
+    render json: @sharing_permission, status: :ok
+    return
```
you don't need a return, ruby will already return the result of the last line, so what is returned by render would be already be returned.  The way it is now it will return nil.
So you can just remove the return here.
```suggestion

```

eramirez405 replied:
Thank you for the feedback.

thiagosil commented on app/javascript/app/models/Document.ts:
```diff
@@ -4,5 +4,5 @@ export type Document = {
     name: string;
     created_at?: string;
-    sharedWith?: User[];
+    sharing_permissions?: User[];
```
Should sharing permissions be a User[] or a SharingPermission[]?

eramirez405 replied:
In the way I prepared the formatted response was to return an array of Users, as the sharing_permission item I use it to get the user information and keep the access_level, something I acknowldge on the solution comment was that sharedWith was a more accurate name, but as I started working on the backend did not see that hint was there, but would definetely change it.

thiagosil commented:
> ### Solution
> To enable users to share Projects, Reports, and Documents while assigning permission levels, I introduced a SharingPermissions table as a control layer. This table includes:
> 
> * resource_type – A string that can be 'project', 'document', or 'report'.
> * access_level – Can be one of 'view', 'edit', or 'full access'.
> 
> Foreign keys:
> 
> * user_id
> * creator_id
> * project_id (if resource_type is 'project')
> * document_id (if resource_type is 'document')
> * report_id (if resource_type is 'report')
> 
> These constraints ensure that each row corresponds to exactly one type of shared resource.
> 
> This solution would allow users to be able to assign projects, documents and reports to other users, and also be able to query for projects, documents, reports assigned to a certain user, documents and reports. Hence creating logic in the server to only allow the edit, index, show controller actions to either the owner/current_user or a user that is accessing a shared resource having the correct access_level.
> 
> Example of projects payload with sharing_permissions included:
> 
> `[ { "id": 3, "name": "Test user project", "description": "test description", "user_id": 5, "created_at": "2025-03-09T23:17:51.498Z", "updated_at": "2025-03-09T23:17:51.498Z", "sharing_permissions": [ { "id": 4, "name": "Robert Paulson", "email": "robertpaulson@example.com", "avatar_url": "https://avatar.iran.liara.run/username?username=robert+paulson", "permission": "view" }, { "id": 1, "name": "Tyler Durden", "email": "tylerdurden@example.com", "avatar_url": "https://avatar.iran.liara.run/username?username=tyler+durden", "permission": "view" } ], "url": "http://localhost:5000/projects/3.json" } ];`
> 
> Ideally I would just return the sharing_permissions to the owner, or users with full access, other than that for shared resources I would just return the access_level the user making the request has, but I struggled with formatting the response in rails, for this reason I used logic in the frontend to identify the access_level of the current_user. Secured the backend and only displayed the options on the frontend based on the permission.
> 
> ### Technical Observations
> * React Hot Reload Missing – Noticed that the React code doesn’t hot reload, which slows development.
> * Formatting Issues – No .prettierrc file; PRs would be easier to review with consistent formatting. I disabled my vscode formatter, but some files got formatted before I realized.
> * Unique Constraints – Initially added unique constraints in a migration, but the server crashed without clear feedback. Rolling back the migration resolved the issue.
> * By default the feature to add a permission was intended to just type the email and proceed, but I changed the user interface to make it more intuitive a user can only type emails of existing users, if having more time I would create a component to stay in the @radix-ui scope, but for simplicity I went with my known friend react-select
> * The user experience of add/update permission of a project from the homepage vs from the project specific page is different, the one in the homepage the ideal more dynamic, in the project specific page I had to close the modal to trigger a refresh on the assigned permissions, with more time the ideal solution would have been to move the logic around assigning permissions to a hook as there is repetitive logic in some places with pretty similar patterns. With that the UX would be consistent for both cases.
> * I started working on the backend and chose the name sharing_permissions then saw on the frontend there was already a suggested property called sharedWith, I think sharedWith is a better name, with more time I would have made changes to use the sharedWith name instead.
> 
> ### Final Thoughts
> This was my first project using Ruby on Rails. My preparation for this take home test was [Ruby in Twenty Minutes](https://www.ruby-lang.org/en/documentation/quickstart/) and then [Getting Started with Rails](https://edgeguides.rubyonrails.org/getting_started.html), they allowed me to understand the soul of ruby on rails in matters of taking away the boilerplate and persisting standard best practices for most of the common server features.
> 
> I got help of chatgpt that help me with some methods that the getting started tutorial did not cover for (e.g., .includes(:project).map(&:project).compact)
> 
> ### My personal to do
> _This is my personal list on how I start tackling the issue by aiming for small tasks—the ones I accomplished and the ones that are next in my soon-to-do scope._
> 
> * [x]  Be able to create permission for project
> * [x]  Be able to retrieve permissions for a project
> * [x]  Create constraint against duplication of permission for project
> * [x]  Be able to create permissions for document
> * [x]  Be able to create permissions for report
> * [x]  Be able to delete permissions on projects
> * [x]  Be able to delete permissions on documents
> * [x]  Be able to delete permissions on reports
> * [x]  Projects can be viewed by owner and users with sharing permission of view, edit or full
> * [x]  Create frontend view for shared projects
> * [x]  Secure in the server sharing features to correct access levels
> * [x]  Show/Hide in front end sharing features to correct access level
> * [x]  Show/Hide add document/report for own/shared
> * [ ]  Show shared documents in the home page
> * [ ]  Show shared reports in the home page
> * [ ]  Secure in server on users with full access and edit can edit projects
> * [ ]  Secure in server on users with full access and edit can edit documents
> * [ ]  Secure in server on users with full access and edit can edit reports
> * [ ]  Display edit button only to owned resources or users with permission in shared resources
> * [ ]  Display the owner of shared resources in the home page
> * [ ]  Clean up frontend current user prop, retrieve from server the access_level of resources shared
> * [ ]  Create components to be shared related to document/report item display, available users (less lines of code, components more readable).
> * [ ]  Separate Project components into different files, for more readability and easier to test independent components if there is more abstraction.
> * [ ]  Explore how to add unit tests for the react frontend with the rails file structure
> * [ ]  Add tests for the backend new features

Let's say we start adding more elements to projects, like issues, files. That also could be shared, would that affect the solution you created?

eramirez405 commented:
> > ### Solution
> > To enable users to share Projects, Reports, and Documents while assigning permission levels, I introduced a SharingPermissions table as a control layer. This table includes:
> > 
> > * resource_type – A string that can be 'project', 'document', or 'report'.
> > * access_level – Can be one of 'view', 'edit', or 'full access'.
> > 
> > Foreign keys:
> > 
> > * user_id
> > * creator_id
> > * project_id (if resource_type is 'project')
> > * document_id (if resource_type is 'document')
> > * report_id (if resource_type is 'report')
> > 
> > These constraints ensure that each row corresponds to exactly one type of shared resource.
> > This solution would allow users to be able to assign projects, documents and reports to other users, and also be able to query for projects, documents, reports assigned to a certain user, documents and reports. Hence creating logic in the server to only allow the edit, index, show controller actions to either the owner/current_user or a user that is accessing a shared resource having the correct access_level.
> > Example of projects payload with sharing_permissions included:
> > `[ { "id": 3, "name": "Test user project", "description": "test description", "user_id": 5, "created_at": "2025-03-09T23:17:51.498Z", "updated_at": "2025-03-09T23:17:51.498Z", "sharing_permissions": [ { "id": 4, "name": "Robert Paulson", "email": "robertpaulson@example.com", "avatar_url": "https://avatar.iran.liara.run/username?username=robert+paulson", "permission": "view" }, { "id": 1, "name": "Tyler Durden", "email": "tylerdurden@example.com", "avatar_url": "https://avatar.iran.liara.run/username?username=tyler+durden", "permission": "view" } ], "url": "http://localhost:5000/projects/3.json" } ];`
> > Ideally I would just return the sharing_permissions to the owner, or users with full access, other than that for shared resources I would just return the access_level the user making the request has, but I struggled with formatting the response in rails, for this reason I used logic in the frontend to identify the access_level of the current_user. Secured the backend and only displayed the options on the frontend based on the permission.
> > ### Technical Observations
> > 
> > * React Hot Reload Missing – Noticed that the React code doesn’t hot reload, which slows development.
> > * Formatting Issues – No .prettierrc file; PRs would be easier to review with consistent formatting. I disabled my vscode formatter, but some files got formatted before I realized.
> > * Unique Constraints – Initially added unique constraints in a migration, but the server crashed without clear feedback. Rolling back the migration resolved the issue.
> > * By default the feature to add a permission was intended to just type the email and proceed, but I changed the user interface to make it more intuitive a user can only type emails of existing users, if having more time I would create a component to stay in the @radix-ui scope, but for simplicity I went with my known friend react-select
> > * The user experience of add/update permission of a project from the homepage vs from the project specific page is different, the one in the homepage the ideal more dynamic, in the project specific page I had to close the modal to trigger a refresh on the assigned permissions, with more time the ideal solution would have been to move the logic around assigning permissions to a hook as there is repetitive logic in some places with pretty similar patterns. With that the UX would be consistent for both cases.
> > * I started working on the backend and chose the name sharing_permissions then saw on the frontend there was already a suggested property called sharedWith, I think sharedWith is a better name, with more time I would have made changes to use the sharedWith name instead.
> > 
> > ### Final Thoughts
> > This was my first project using Ruby on Rails. My preparation for this take home test was [Ruby in Twenty Minutes](https://www.ruby-lang.org/en/documentation/quickstart/) and then [Getting Started with Rails](https://edgeguides.rubyonrails.org/getting_started.html), they allowed me to understand the soul of ruby on rails in matters of taking away the boilerplate and persisting standard best practices for most of the common server features.
> > I got help of chatgpt that help me with some methods that the getting started tutorial did not cover for (e.g., .includes(:project).map(&:project).compact)
> > ### My personal to do
> > _This is my personal list on how I start tackling the issue by aiming for small tasks—the ones I accomplished and the ones that are next in my soon-to-do scope._
> > 
> > * [x]   Be able to create permission for project
> > * [x]   Be able to retrieve permissions for a project
> > * [x]   Create constraint against duplication of permission for project
> > * [x]   Be able to create permissions for document
> > * [x]   Be able to create permissions for report
> > * [x]   Be able to delete permissions on projects
> > * [x]   Be able to delete permissions on documents
> > * [x]   Be able to delete permissions on reports
> > * [x]   Projects can be viewed by owner and users with sharing permission of view, edit or full
> > * [x]   Create frontend view for shared projects
> > * [x]   Secure in the server sharing features to correct access levels
> > * [x]   Show/Hide in front end sharing features to correct access level
> > * [x]   Show/Hide add document/report for own/shared
> > * [ ]   Show shared documents in the home page
> > * [ ]   Show shared reports in the home page
> > * [ ]   Secure in server on users with full access and edit can edit projects
> > * [ ]   Secure in server on users with full access and edit can edit documents
> > * [ ]   Secure in server on users with full access and edit can edit reports
> > * [ ]   Display edit button only to owned resources or users with permission in shared resources
> > * [ ]   Display the owner of shared resources in the home page
> > * [ ]   Clean up frontend current user prop, retrieve from server the access_level of resources shared
> > * [ ]   Create components to be shared related to document/report item display, available users (less lines of code, components more readable).
> > * [ ]   Separate Project components into different files, for more readability and easier to test independent components if there is more abstraction.
> > * [ ]   Explore how to add unit tests for the react frontend with the rails file structure
> > * [ ]   Add tests for the backend new features
> 
> Let's say we start adding more elements to projects, like issues, files. That also could be shared, would that affect the solution you created?

The solution is flexible and prepared to stay within the sharing_permissions scope, if for instance we added "issues" to belong to "projects" and be shareable, once the routes, controller, view and model for issue is created, we would only need to:

- mark in the issue model that an issue "has_many :sharing_permissions" and "belongs_to :project"
- create a db migration to include the issue_id column
- update the sharing_permission validation for resource_type so it also accepts "issue" as acceptable value
- mark in the sharing_permission model that it "belongs_to :issue, optional: true"
- update the sharing_permission model so the validators include rules to ensure issue_id is present on the right circunstances
- update the sharing_permission controller actions so the switch logic also contemplates the "issue" case
- use a similar approach as in documents views for the issues view so the sharing_permissions are part of the payload
